src/pykeen/sampling/basic_negative_sampler.py

```diff
@@ -2,60 +2,122 @@
 
 """Negative sampling algorithm based on the work of of Bordes *et al.*."""
 
+from typing import Collection, Optional, Tuple
+
 import torch
 
 from .negative_sampler import NegativeSampler
+from ..triples import TriplesFactory
 
 __all__ = [
     'BasicNegativeSampler',
 ]
 
+LOOKUP = {'h': 0, 'r': 1, 't': 2}
+
 
 class BasicNegativeSampler(NegativeSampler):
     r"""A basic negative sampler.
 
-    This negative sampler that corrupts positive triples $(h,r,t) \in \mathcal{K}$ by replacing either $h$ or $t$.
+    This negative sampler that corrupts positive triples $(h,r,t) \in \mathcal{K}$ by replacing either $h$, $r$ or $t$
+    based on the chosen corruption scheme. The corruption scheme can contain $h$, $r$ and $t$ or any subset of these.
 
     Steps:
 
-    1. Randomly (uniformly) determine whether $h$ or $t$ shall be corrupted for a positive triple
+    1. Randomly (uniformly) determine whether $h$, $r$ or $t$ shall be corrupted for a positive triple
        $(h,r,t) \in \mathcal{K}$.
-    2. Randomly (uniformly) sample an entity $e \in \mathcal{E}$ for selection to corrupt the triple.
+    2. Randomly (uniformly) sample an entity $e \in \mathcal{E}$ or relation $r' \in \mathcal{R}$ for selection to
+       corrupt the triple.
 
        - If $h$ was selected before, the corrupted triple is $(e,r,t)$
+       - If $r$ was selected before, the corrupted triple is $(h,r',t)$
        - If $t$ was selected before, the corrupted triple is $(h,r,e)$
+    3. If ``filtered`` is set to ``True``, all proposed corrupted triples that also exist as
+       actual positive triples $(h,r,t) \in \mathcal{K}$ will be removed.
     """
 
     #: The default strategy for optimizing the negative sampler's hyper-parameters
     hpo_default = dict(
         num_negs_per_pos=dict(type=int, low=1, high=100, q=10),
     )
 
-    def sample(self, positive_batch: torch.LongTensor) -> torch.LongTensor:
+    def __init__(
+        self,
+        triples_factory: TriplesFactory,
+        num_negs_per_pos: Optional[int] = None,
+        filtered: bool = False,
+        corruption_scheme: Optional[Collection[str]] = None,
+    ) -> None:
+        """Initialize the negative sampler with the given entities.
+
+        :param triples_factory: The factory holding the triples to sample from
+        :param num_negs_per_pos: Number of negative samples to make per positive triple. Defaults to 1.
+        :param filtered: Whether proposed corrupted triples that are in the training data should be filtered.
+            Defaults to False.
```

Comment:
Add an additional comment / warning to this parameter that this corruption scheme is significantly more expensive, in particular for large number of negative examples. And maybe even add a comment that depending on the dataset the likelihood of generating a false negative is small (last one is optional).

Reply:
I have seen below that an explanation is provided in `filter_negative_triples`. Also link this method here?